xrpl/models/requests/generic_request.py

```diff
@@ -62,7 +63,16 @@ def from_dict(cls: Type[GenericRequest], value: Dict[str, Any]) -> GenericReques
 
         elif "method" in value:  # JSON RPC formatting
             if "params" in value:  # actual JSON RPC formatting
-                value = {"method": value["method"], **value["params"]}
+                if "api_version" in value:
```

Comment:
Why does the API version need to be specified in generic requests specifically?

Reply:
Certain requests like `tx_history` are implemented through `GenericRequest`. But this request ahs been deprecated since api version 2. This PR intends to make API v2 as the default response in xrpl-py.

An `api_version < 2` needs to be specified for certain generic-requests, otherwise they fail with an `invalid_API_version` error.

Look at `test_generic_request` for an example

Reply:
Seems fine to leave that as existing behavior then - default to v2. For example, `GenericRequest` is also used for `ledger_accept`, which doesn't differ between V1 and V2.

Reply:
Yes, but `api_version: 2` is specified in all requests, by default (including `ledger_accept`).

Without this change, `tx_history` requests are transmitted with `api_version: 2` and they fail. Since we need to support API v1, how do you propose to handle `tx_history` requests?

Reply:
It should error if `api_version` isn't included, since the default is `2`. If `api_version: 1` is included, it will work. I assume that's what would still happen.

Note: we don't have to specifically support `tx_history`, that was just something to use for the test since the RPC doesn't have a model.

Reply:
> It should error if `api_version` isn't included, since the default is `2`. If `api_version: 1` is included, it will work. I assume that's what would still happen.
> 

No, `api_version: 2` is always included in every request, because `Request` is the parent class of every Request. By default, this is the Request JSON:

```
{
    "method": "tx_history",
    "params": {
        "start": 0,
    },
    "api_version": 2, // we need to change this value into 1
}
```

Reply:
> Note: we don't have to specifically support `tx_history`, that was just something to use for the test since the RPC doesn't have a model.

I think `tx_history` is a good example because it does not work in rippled API v2 (or later versions). This change "forwards" the `api_version` as specified by the user.

Reply:
Agree with @mvadari - it's ok for `tx_history` fail now that we default to v2.

Reply:
> No, `api_version: 2` is always included in every request, because `Request` is the parent class of every Request.

Yes, I understand. I would expect this to fail:
```python
GenericRequest(method="tx_history", start=0) # api_version defaults to 2
```

But this to succeed:
```python
GenericRequest(method="tx_history", start=0, api_version=1)
```

Reply:
> But this to succeed:
> 
> ```python
> GenericRequest(method="tx_history", start=0, api_version=1)
> ```
> 

This input will not succeed, because we are over-writing the `value` variable, hence we lose the `api_version` information even if it specified.

In the below line, `value` variable is overwritten as:
```
value = {
              # "api_version": value["api_version"],
              "method": value["method"],
              **value["params"],
          }
```






> I disagree that this behavior means that any user of `GenericRequest` needs to specify an API version, since most APIs did not change from API v1 to v2.

I'm not forcing users to specify any version at all. I'm preserving the `api_version` information from being lost.

Reply:
You don't need the API check for that. I tried running just this code in the test and it worked:
```
if "params" in value:  # actual JSON RPC formatting
  value = {
    "api_version": value["api_version"],
    "method": value["method"],
    **value["params"],
  }
```

Reply:
what if a user submits a raw JSON input, but forgets to specify the `api_version`?

Since the user is not using the `GenericRequest/Request` class, the `api_version` field will need to be specified manually

Reply:
The user can't submit raw JSON input in xrpl-py, you have to go through a model.

Even if you could, it would then default to whatever the rippled instance has as default (probably v1).

Reply:
I added a test case at the end of this file: `tests/integration/reqs/test_generic_request.py`

Users can submit raw-JSON inputs through the `from_dict` methods. This request does not reach rippled due to processing errors in the `from_dict` function.

We will need to document that `api_version` is a mandatory field, if raw-JSON input is provided. If `GenericRequest` model is used, then the default `api_version` is provided automatically.

Reply:
It's not a mandatory field if a default option is provided. The same is true of any other API method - if you don't specify an `api_version` field, you'll get the default version.